### PR TITLE
Use a PodTemplate resource in the Helm Chart

### PR DESCRIPTION
**Pull request by swiatekm**
## What does this PR do?

Instead of an untyped .tpl Helm template, use a PodTemplate resource. This resource is loaded from a file, templated via the tpl function, converted to a dict to get a specific key, and then converted back to YAML.

This is a bit convoluted, but it gives us intellisense for the pod template.

## Why is it important?

.tpl files aren't very readable. On the other hand, Kubernetes resources get pretty good support from IDEs and tools. So this is a maintainability improvement.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Review by pkoutsovasilis — COMMENTED** [review]
ok now I see what you meant back in the day during the initial Helm chart PR, a couple of quick ones:
* Still neither Jetbrain-based IDEs (Kubernetes plugin installed) nor VSCode (Kubernetes plugin installed) give me intellisense hints about `pod_template.yaml`; do I need a specific configuration for this to work?
* Not a super fan that this needs to live inside a conf folder but I understand why you went down this path; however, couldn't you pass a fixed variable from the tpl invocation and have this file rendered only when invoked like that?! I believe that being able to keep the files that regard elastic-agent for k8s/eck rendering under a single path makes it easier during development
* we want to do the same for ECK right?!

**Comment by swiatekm** [comment]
> * Still neither Jetbrain-based IDEs (Kubernetes plugin installed) nor VSCode (Kubernetes plugin installed) give me intellisense hints about `pod_template.yaml`; do I need a specific configuration for this to work?

It doesn't work by default because both of these only treat files in the Chart's `templates` directory as Helm templates. In Goland, you need to go to `Settings` -> `Editor` -> `File Types` and add the file to the detection pattern. In VS Code you can just change the type to `helm-template` directly. There's also some global setting for it, but I don't recall what it is off the top of my head.

Putting that aside, having yaml syntax highlighting in Github is nice in and out of itself.

> * Not a super fan that this needs to live inside a conf folder but I understand why you went down this path; however, couldn't you pass a fixed variable from the tpl invocation and have this file rendered only when invoked like that?! I believe that being able to keep the files that regard elastic-agent for k8s/eck rendering under a single path makes it easier during development

Unfortunately, [Helm doesn't allow us](https://helm.sh/docs/chart_template_guide/accessing_files/) to access files under `templates/` via `.Files`. Though I'm very open to renaming this folder, no strong feelings there.

> * we want to do the same for ECK right?!

If we decide to do it here, then probably, yes.

**Comment by pkoutsovasilis** [comment]
I think there is a less convoluted way to achieve the same result you are after; I have tested this locally, but try it yourself:
1. revert your changes
2. rename `k8s/_pod_template.tpl` to `k8s/_pod_template.yaml`
3. add in `k8s/_pod_template.yaml` after `{{- $agentName := index . 2 }}` the following
  ```
  apiVersion: v1
  kind: PodTemplate
  template:
    spec: ...
  ```
4. replace in all `k8s/{daemoset,deployment,statefulset.yaml}`, the following line
```
{{ include "elasticagent.engine.k8s.podTemplate" (list $ $presetVal $agentName) }} 
```
with
```
{{- $podTemplateResource := include "elasticagent.engine.k8s.podTemplate" (list $ $presetVal $agentName) | fromYaml  -}}
{{- toYaml ($podTemplateResource).template }}
```

**Comment by swiatekm** [comment]
Huh, I thought only .tpl files could have named templates defined in them. If that works, it'd be way better indeed.

**Comment by swiatekm** [comment]
@pkoutsovasilis how about this? I added a new named template in `_pod_template.yaml`, which contains the resource, and which the existing template calls. Actual manifests just need to indent it appropriately.

**Comment by pkoutsovasilis** [comment]
> @pkoutsovasilis how about this? I added a new named template in `_pod_template.yaml`, which contains the resource, and which the existing template calls. Actual manifests just need to indent it appropriately.

any particular reason you want to avoid following my sequence [here](https://github.com/elastic/elastic-agent/pull/5831#issuecomment-2429096246)? Instead you went for an extra named template, but I was thinking that we want reduce them?!

**Comment by swiatekm** [comment]
> > @pkoutsovasilis how about this? I added a new named template in `_pod_template.yaml`, which contains the resource, and which the existing template calls. Actual manifests just need to indent it appropriately.
> 
> any particular reason you want to avoid following my sequence [here](https://github.com/elastic/elastic-agent/pull/5831#issuecomment-2429096246) (why we should keep `_pod_template.tpl`)? Instead you went for an extra named template, but I was thinking that we want reduce them or at least not increase them?!

I thought I would do the `toYaml` and `fromYaml` song and dance in the named template so it doesn't need to happen at each call site. But it's a minor thing, if you don't like it, I can do what you proposed.

**Comment by swiatekm** [comment]
@pkoutsovasilis I've made the change you requested, does it look right to you now?

**Comment by pkoutsovasilis** [comment]
> @pkoutsovasilis I've made the change you requested, does it look right to you now?

It does, but I am wondering if we have to change the order of the yaml fields inside the `_pod_template.yaml`?! this makes unnecessary additions and removal shown as diffs. As a result it is harder now to track down what is actually changed and what is not. Can you just rename `_pod_template.tpl` to `_pod_template.yaml` and then just add anything extra you need?! Also the same needs to go for `eck`, ty 🙂

**Comment by swiatekm** [comment]
> > @pkoutsovasilis I've made the change you requested, does it look right to you now?
> 
> It does, but I am wondering if we have to change the order of the yaml fields inside the `_pod_template.yaml`?! this makes unnecessary additions and removal shown as diffs. As a result it is harder now to track down what is actually changed and what is not. Can you just rename `_pod_template.tpl` to `_pod_template.yaml` and then just add anything extra you need?! Also the same needs to go for `eck`, ty 🙂

I haven't changed the order. This is a side effect of unmarshalling the template from yaml, and then marshalling it back. This operation doesn't preserve key order, so we get them in alphabetical order at the end.

Will do for `eck`.

**Comment by swiatekm** [comment]
@pkoutsovasilis modified the ECK pod template as well.

File: deploy/helm/elastic-agent/templates/agent/eck/deployment.yaml [hunk 1]
```diff
@@ -37,7 +38,7 @@ spec:
     replicas: {{ . }}
     {{- end }}
     podTemplate:
-{{ include "elasticagent.engine.eck.podTemplate" (list $ $presetVal $agentName) }}
+{{- toYaml ($podTemplateResource).template | nindent 4 }}
```
Comment (pkoutsovasilis) [inline deploy/helm/elastic-agent/templates/agent/eck/deployment.yaml:41]:
```suggestion
{{- toYaml ($podTemplateResource).template | nindent 6 }}
```

File: deploy/helm/elastic-agent/templates/agent/eck/statefulset.yaml [hunk 2]
```diff
@@ -37,7 +38,7 @@ spec:
     replicas: {{ . }}
     {{- end }}
     podTemplate:
-{{ include "elasticagent.engine.eck.podTemplate" (list $ $presetVal $agentName) }}
+{{- toYaml ($podTemplateResource).template | nindent 4 }}
```
Comment (pkoutsovasilis) [inline deploy/helm/elastic-agent/templates/agent/eck/statefulset.yaml:41]:
```suggestion
{{- toYaml ($podTemplateResource).template | nindent 6 }}
```
`spec` for eck should be under `podTemplate`

File: deploy/helm/elastic-agent/examples/kubernetes-default/rendered/manifest.yaml [hunk 3]
```diff
@@ -1291,102 +1292,103 @@ spec:
...
-                  fieldPath: metadata.name
-            - name: STATE_PATH
-              value: "/usr/share/elastic-agent/state"
-            - name: ELASTIC_NETINFO
-              value: "false"
+      - emptyDir: {}
+        name: agent-data
+      - name: config
+        secret:
+          defaultMode: 292
```
[the remark this answers is not in the document]
Comment (swiatekm) [reply deploy/helm/elastic-agent/examples/kubernetes-default/rendered/manifest.yaml:1393]: Yeah, this is another yaml marshalling thing. In Yaml, `0444` is just a number, and by default the Helm marshalling outputs base 10.

File: deploy/helm/elastic-agent/templates/agent/eck/daemonset.yaml [hunk 4]
```diff
@@ -34,7 +35,7 @@ spec:
   {{- end }}
   daemonSet:
     podTemplate:
-{{ include "elasticagent.engine.eck.podTemplate" (list $ $presetVal $agentName) }}
+{{- toYaml ($podTemplateResource).template | nindent 4 }}
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply deploy/helm/elastic-agent/templates/agent/eck/daemonset.yaml]: yep that sounds like a good idea, we should have a follow up on that 🙂

**Review by pkoutsovasilis — APPROVED** [review]
LGTM. Proper hints in helm templates, nice stuff 🙂

File: deploy/helm/elastic-agent/templates/agent/eck/daemonset.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ycombinator) [reply deploy/helm/elastic-agent/templates/agent/eck/daemonset.yaml]: @swiatekm Would you mind making an issue for the follow up Go test you mentioned above? Thanks.

[the remark this answers is not in the document]
Comment (swiatekm) [reply deploy/helm/elastic-agent/templates/agent/eck/daemonset.yaml]: https://github.com/elastic/elastic-agent/issues/6014

**Pull request merged**